db/migrate/20171222120314_add_constraints_on_discovery_rules_hostgroups.rb

```diff
@@ -0,0 +1,14 @@
+ class AddConstraintsOnDiscoveryRulesHostgroups < ActiveRecord::Migration[4.2]
+  
+   def change
+    execute "DELETE from discovery_rules dr where NOT EXISTS (select NULL from hostgroups hg where hg.id = dr.hostgroup_id);"
```

Comment:
Not exactly sure what you are trying to do here, but this should be:

    DiscoveryRule.where("hostgroup_id IS NULL").delete_all

Reply:
Here, I have checked in both the tables - `discovery_rules` and `hostgroups` table. I have checked whether the hostgroup_id(which is a column in the discovery_rules table) has an equivalent id(which is a column of the hostgroups table). 
I have tested this query and it works really efficiently :)

The solution that you have provided in the review, say if there is a discovery rule present but the hostgroup is deleted, then, in that case, you will face a `FOREIGNKEY_VIOLATION` error, to avoid that I suggest we use the above query.

Reply:
Ah in that case I am in favour of:

    DiscoveryRule.where("hostgroup_id IS NULL OR hostgroup_id NOT IN (?)", Hostgroup.all.pluck(:id)).delete_all

Your query is more efficient but we are dealing with hundreds of records maximum. MySQL does not support subselects and I want to avoid testing this on MySQL to be honest (my tip: won't work). Thanks :-)

Reply:
Correction!

    DiscoveryRule.unscoped.where("hostgroup_id IS NULL OR hostgroup_id NOT IN (?)", Hostgroup.unscoped.all.pluck(:id)).delete_all

Reply:
@lzap done.